File: server/store.ts
import { randomUUID, createHash } from "node:crypto";
import fs from "node:fs/promises";
import os from "node:os";
import path from "node:path";
import type { AppSettings, OperationRecord, RepoRecord } from "../src/types";

type AppState = {
  repos: RepoRecord[];
  operations: OperationRecord[];
  settings: AppSettings;
};

const defaultState: AppState = {
  repos: [],
  operations: [],
  settings: {
    safeMode: true,
    integrations: {
      editor: "auto",
      terminal: "auto"
    }
  }
};

export class AppStore {
  private writeQueue: Promise<void> = Promise.resolve();

  constructor(
    private readonly stateFile = path.join(process.cwd(), ".worktree-manager", "state.json"),
    private readonly fallbackStateFiles: string[] = []
  ) {}

  stateFilePath(): string {
    return this.stateFile;
  }

  async listRepos(): Promise<RepoRecord[]> {
    const state = await this.readState();
    return state.repos;
  }

  async getRepo(repoId: string): Promise<RepoRecord | null> {
    const state = await this.readState();
    return state.repos.find((repo) => repo.id === repoId) ?? null;
  }

  async upsertRepo(repoPath: string): Promise<RepoRecord> {
    const resolvedPath = path.resolve(repoPath);
    const repo: RepoRecord = {
      id: createHash("sha1").update(resolvedPath).digest("hex").slice(0, 16),
      name: path.basename(resolvedPath),
      path: resolvedPath,
      lastOpenedAt: new Date().toISOString()
    };

    return this.updateState((state) => {
      state.repos = [repo, ...state.repos.filter((item) => item.id !== repo.id)].slice(0, 12);
      return repo;
    });
  }

  async getSettings(): Promise<AppSettings> {
    const state = await this.readState();
    return state.settings;
  }

  async updateSettings(settings: Partial<AppSettings>): Promise<AppSettings> {
    return this.updateState((state) => {
      state.settings = normalizeSettings({ ...state.settings, ...settings });
      return state.settings;
    });
  }

  async recordOperation(
    operation: Omit<OperationRecord, "id" | "finishedAt">
  ): Promise<OperationRecord> {
    const record: OperationRecord = {
      ...operation,
      id: randomUUID(),
      finishedAt: new Date().toISOString()
    };

    return this.updateState((state) => {
      state.operations = [record, ...state.operations].slice(0, 60);
      return record;
    });
  }

  async listOperations(): Promise<OperationRecord[]> {
    const state = await this.readState();
    return state.operations;
  }

  async getOperation(operationId: string): Promise<OperationRecord | null> {
    const state = await this.readState();
    return state.operations.find((operation) => operation.id === operationId) ?? null;
  }

  private async readState(): Promise<AppState> {
    try {
      return parseState(await fs.readFile(this.stateFile, "utf8"));
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code === "ENOENT") {
        for (const fallbackFile of this.fallbackStateFiles) {
          try {
            const state = parseState(await fs.readFile(fallbackFile, "utf8"));
            await this.writeState(state);
            return state;
          } catch (fallbackError) {
            if ((fallbackError as NodeJS.ErrnoException).code !== "ENOENT") {
              throw fallbackError;
            }
          }
        }

        return { ...defaultState };
      }

      throw error;
    }
  }

  private async writeState(state: AppState): Promise<void> {
    await fs.mkdir(path.dirname(this.stateFile), { recursive: true });
    const tempFile = `${this.stateFile}.${process.pid}.${Date.now()}.tmp`;
    await fs.writeFile(tempFile, JSON.stringify(state, null, 2));
    await fs.rename(tempFile, this.stateFile);
  }

  private async updateState<T>(mutator: (state: AppState) => T): Promise<T> {
    let result: T;
    const update = this.writeQueue.then(async () => {
      const state = await this.readState();
      result = mutator(state);
      await this.writeState(state);
    });

    this.writeQueue = update.then(
      () => undefined,
      () => undefined
    );
    await update;
    return result!;
  }
}

export function createDefaultStore(): AppStore {
  return new AppStore(defaultStateFile(), [legacyStateFile()]);
}

function parseState(contents: string): AppState {
  const parsed = JSON.parse(contents) as AppState;
  return {
    repos: Array.isArray(parsed.repos) ? parsed.repos : [],
    operations: Array.isArray(parsed.operations) ? parsed.operations : [],
    settings: normalizeSettings(parsed.settings)
  };
}

function normalizeSettings(value: unknown): AppSettings {
  const settings = value as Partial<AppSettings> | null;
  return {
    safeMode: typeof settings?.safeMode === "boolean" ? settings.safeMode : true,
    integrations: {
      editor: isEditorIntegration(settings?.integrations?.editor)
        ? settings.integrations.editor
        : "auto",
      terminal: isTerminalIntegration(settings?.integrations?.terminal)
        ? settings.integrations.terminal
        : "auto"
    }
  };
}

function isEditorIntegration(value: unknown): value is AppSettings["integrations"]["editor"] {
  return (
    value === "auto" ||
    value === "vscode" ||
    value === "cursor" ||
    value === "windsurf" ||
    value === "zed" ||
    value === "sublime"
  );
}

function isTerminalIntegration(value: unknown): value is AppSettings["integrations"]["terminal"] {
  return (
    value === "auto" ||
    value === "system" ||
    value === "iterm" ||
    value === "warp" ||
    value === "windows-terminal" ||
    value === "x-terminal-emulator" ||
    value === "gnome-terminal" ||
    value === "konsole"
  );
}

function defaultStateFile(): string {
  const configuredDir = process.env.WORKTREE_MANAGER_STATE_DIR;
  if (configuredDir) return path.join(configuredDir, "state.json");

  if (process.platform === "darwin") {
    return path.join(os.homedir(), "Library", "Application Support", "Worktree Manager", "state.json");
  }

  if (process.platform === "win32") {
    return path.join(
      process.env.APPDATA ?? path.join(os.homedir(), "AppData", "Roaming"),
      "Worktree Manager",
      "state.json"
    );
  }

  return path.join(
    process.env.XDG_STATE_HOME ?? path.join(os.homedir(), ".local", "state"),
    "worktree-manager",
    "state.json"
  );
}

function legacyStateFile(): string {
  return path.join(process.cwd(), ".worktree-manager", "state.json");
}

File: src-tauri/src/store.rs
use crate::models::{AppSettings, AppSettingsPatch, OperationRecord, RepoRecord};
use chrono::{SecondsFormat, Utc};
use serde::{Deserialize, Serialize};
use sha1::{Digest, Sha1};
use std::{
    env, fs,
    path::{Path, PathBuf},
    sync::Mutex,
};
use uuid::Uuid;

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
struct StoreData {
    repos: Vec<RepoRecord>,
    operations: Vec<OperationRecord>,
    #[serde(default)]
    settings: AppSettings,
}

pub struct AppState {
    state_file: PathBuf,
    data: Mutex<StoreData>,
    pub git_lock: Mutex<()>,
}

impl AppState {
    pub fn load() -> Self {
        let state_file = default_state_file();
        let data = read_state(&state_file)
            .or_else(|| {
                let legacy = legacy_state_file();
                let state = read_state(&legacy)?;
                let _ = write_state(&state_file, &state);
                Some(state)
            })
            .unwrap_or_default();

        Self {
            state_file,
            data: Mutex::new(data),
            git_lock: Mutex::new(()),
        }
    }

    pub fn state_file_path(&self) -> String {
        path_string(&self.state_file)
    }

    pub fn list_repos(&self) -> Vec<RepoRecord> {
        self.data
            .lock()
            .expect("store mutex poisoned")
            .repos
            .clone()
    }

    pub fn get_repo(&self, repo_id: &str) -> Option<RepoRecord> {
        self.data
            .lock()
            .expect("store mutex poisoned")
            .repos
            .iter()
            .find(|repo| repo.id == repo_id)
            .cloned()
    }

    pub fn upsert_repo(&self, repo_path: &Path) -> Result<RepoRecord, String> {
        let resolved = absolute_path(repo_path);
        let repo = RepoRecord {
            id: repo_id(&resolved),
            name: resolved
                .file_name()
                .map(|value| value.to_string_lossy().to_string())
                .unwrap_or_else(|| "Repositorio".to_string()),
            path: path_string(&resolved),
            last_opened_at: now_iso(),
        };

        let mut data = self.data.lock().expect("store mutex poisoned");
        data.repos.retain(|item| item.id != repo.id);
        data.repos.insert(0, repo.clone());
        data.repos.truncate(12);
        write_state(&self.state_file, &data)?;
        Ok(repo)
    }

    pub fn get_settings(&self) -> AppSettings {
        self.data
            .lock()
            .expect("store mutex poisoned")
            .settings
            .clone()
    }

    pub fn update_settings(&self, patch: AppSettingsPatch) -> Result<AppSettings, String> {
        let mut data = self.data.lock().expect("store mutex poisoned");
        if let Some(safe_mode) = patch.safe_mode {
            data.settings.safe_mode = safe_mode;
        }
        if let Some(locale) = patch.locale {
            data.settings.locale = locale;
        }
        if let Some(integrations) = patch.integrations {
            data.settings.integrations = integrations;
        }
        let settings = data.settings.clone();
        write_state(&self.state_file, &data)?;
        Ok(settings)
    }

    pub fn record_operation(&self, mut operation: OperationRecord) -> OperationRecord {
        operation.id = Uuid::new_v4().to_string();
        operation.finished_at = now_iso();

        let mut data = self.data.lock().expect("store mutex poisoned");
        data.operations.insert(0, operation);
        data.operations.truncate(60);
        let record = data.operations[0].clone();
        let _ = write_state(&self.state_file, &data);
        record
    }

    pub fn list_operations(&self) -> Vec<OperationRecord> {
        self.data
            .lock()
            .expect("store mutex poisoned")
            .operations
            .clone()
    }

    pub fn get_operation(&self, operation_id: &str) -> Option<OperationRecord> {
        self.data
            .lock()
            .expect("store mutex poisoned")
            .operations
            .iter()
            .find(|operation| operation.id == operation_id)
            .cloned()
    }
}

pub fn now_iso() -> String {
    Utc::now().to_rfc3339_opts(SecondsFormat::Millis, true)
}

pub fn absolute_path(path: &Path) -> PathBuf {
    if path.is_absolute() {
        path.to_path_buf()
    } else {
        env::current_dir()
            .unwrap_or_else(|_| PathBuf::from("."))
            .join(path)
    }
}

pub fn path_string(path: &Path) -> String {
    path.to_string_lossy().to_string()
}

fn repo_id(path: &Path) -> String {
    let mut hasher = Sha1::new();
    hasher.update(path_string(path).as_bytes());
    let digest = hasher.finalize();
    digest
        .iter()
        .take(8)
        .map(|byte| format!("{byte:02x}"))
        .collect()
}

fn read_state(path: &Path) -> Option<StoreData> {
    let contents = fs::read_to_string(path).ok()?;
    let parsed = serde_json::from_str::<StoreData>(&contents).ok()?;
    Some(parsed)
}

fn write_state(path: &Path, state: &StoreData) -> Result<(), String> {
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent).map_err(|error| error.to_string())?;
    }

    let temp_file = path.with_extension(format!(
        "json.{}.{}.tmp",
        std::process::id(),
        Utc::now().timestamp_millis()
    ));
    let contents = serde_json::to_string_pretty(state).map_err(|error| error.to_string())?;
    fs::write(&temp_file, contents).map_err(|error| error.to_string())?;
    fs::rename(&temp_file, path).map_err(|error| error.to_string())?;
    Ok(())
}

fn default_state_file() -> PathBuf {
    if let Ok(configured_dir) = env::var("WORKTREE_MANAGER_STATE_DIR") {
        if !configured_dir.trim().is_empty() {
            return PathBuf::from(configured_dir).join("state.json");
        }
    }

    if cfg!(target_os = "macos") {
        return home_dir()
            .join("Library")
            .join("Application Support")
            .join("Worktree Manager")
            .join("state.json");
    }

    if cfg!(target_os = "windows") {
        return PathBuf::from(
            env::var_os("APPDATA")
                .unwrap_or_else(|| home_dir().join("AppData").join("Roaming").into_os_string()),
        )
        .join("Worktree Manager")
        .join("state.json");
    }

    PathBuf::from(
        env::var_os("XDG_STATE_HOME")
            .unwrap_or_else(|| home_dir().join(".local").join("state").into_os_string()),
    )
    .join("worktree-manager")
    .join("state.json")
}

fn legacy_state_file() -> PathBuf {
    env::current_dir()
        .unwrap_or_else(|_| PathBuf::from("."))
        .join(".worktree-manager")
        .join("state.json")
}

fn home_dir() -> PathBuf {
    env::var_os("HOME")
        .map(PathBuf::from)
        .or_else(|| env::var_os("USERPROFILE").map(PathBuf::from))
        .unwrap_or_else(|| PathBuf::from("."))
}
